library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/brewing/CustomPotionBrewingRecipe.java

```diff
@@ -0,0 +1,188 @@
+ *       "duration": 10
+ *     }
+ *   ]
+ * }
+ * </code></pre>
+ * @see PotionUtil#getCustomPotionEffects(ItemStack)
+ */
+public class CustomPotionBrewingRecipe extends SimplePotionBrewingRecipe {
+	private final List<StatusEffectInstance> statusEffects = new ArrayList<>();
+	public CustomPotionBrewingRecipe(Identifier id, String group, Potion input, Ingredient ingredient, Potion output, int fuel, int brewTime) {
+		super(id, group, input, ingredient, output, fuel, brewTime);
+		PotionUtil.setCustomPotionEffects(this.result, this.statusEffects);
+	}
+
+	public CustomPotionBrewingRecipe(Identifier id, String group, Potion input, Ingredient ingredient, Potion output, int fuel, int brewTime, Collection<StatusEffectInstance> effects) {
+		this(id, group, input, ingredient, output, fuel, brewTime);
+		this.statusEffects.addAll(effects);
+		PotionUtil.setCustomPotionEffects(this.result, this.statusEffects);
+	}
+
+	@Override
+	protected ItemStack craft(int slot, ItemStack input) {
+		ItemStack output = super.craft(slot, input);
+
+		var statusEffects = this.statusEffects;
+		if (input.getOrCreateNbt().contains("CustomPotionEffects")) {
+			statusEffects = new ArrayList<>(this.statusEffects);
+			statusEffects.addAll(PotionUtil.getCustomPotionEffects(input));
+		}
+
+		PotionUtil.setCustomPotionEffects(output, statusEffects);
+
+		return output;
+	}
+
+	@Override
+	public RecipeSerializer<CustomPotionBrewingRecipe> getSerializer() {
+		return RecipeImpl.CUSTOM_POTION_SERIALIZER;
+	}
+
+	public static class Serializer extends SimplePotionBrewingRecipe.Serializer<CustomPotionBrewingRecipe> {
+		public Serializer(RecipeFactory<Potion, CustomPotionBrewingRecipe> recipeFactory) {
+			super(recipeFactory);
+		}
+
+		@Override
+		public CustomPotionBrewingRecipe read(Identifier id, JsonObject json) {
+			CustomPotionBrewingRecipe recipe = super.read(id, json);
+
+			if (json.has("effects")) {
+				recipe.statusEffects.clear();
+				for (JsonElement effectJson : JsonHelper.getArray(json, "effects")) {
+					StatusEffectInstance instance;
+					instance = this.readStatusEffectInstance(effectJson);
+					recipe.statusEffects.add(instance);
+				}
+			}
+
+			return recipe;
+		}
+
+		private StatusEffectInstance readStatusEffectInstance(JsonElement json) {
+			if (json instanceof JsonObject jsonObject) {
+				String string = JsonHelper.getString(jsonObject, "type");
+				StatusEffect statusEffect = Registries.STATUS_EFFECT
+						.getOrEmpty(new Identifier(string))
+						.orElseThrow(() -> new JsonSyntaxException("Unknown status effect '" + string + "'"));
+				int duration = JsonHelper.getInt(jsonObject, "duration", 3600);
+				int amplifier = JsonHelper.getInt(jsonObject, "amplifier", 0);
+				boolean showParticles = JsonHelper.getBoolean(jsonObject, "particles", true);
+				boolean showIcon = JsonHelper.getBoolean(jsonObject, "icon", true);
+				return new StatusEffectInstance(statusEffect, duration, amplifier, false, showParticles, showIcon);
```

Comment:
Follow the Checkstyle, have new line before `return`s!
```suggestion
				boolean showIcon = JsonHelper.getBoolean(jsonObject, "icon", true);

				return new StatusEffectInstance(statusEffect, duration, amplifier, false, showParticles, showIcon);
```